In [1]:
bw(s,msq,Γ)=1/(msq-s-1.0im*sqrt(msq)*Γ)

bw (generic function with 1 method)

In [2]:
Pkg.add("Plots")
Pkg.add("DataFrames")
Pkg.add("StatsBase")
Pkg.add("QuadGK")
Pkg.add("GR")

INFO: Package Plots is already installed
INFO: METADATA is out-of-date — you may not have the latest version of Plots
INFO: Use `Pkg.update()` to get the latest versions of your packages
INFO: Package DataFrames is already installed
INFO: METADATA is out-of-date — you may not have the latest version of DataFrames
INFO: Use `Pkg.update()` to get the latest versions of your packages
INFO: Package StatsBase is already installed
INFO: METADATA is out-of-date — you may not have the latest version of StatsBase
INFO: Use `Pkg.update()` to get the latest versions of your packages
INFO: Package QuadGK is already installed
INFO: METADATA is out-of-date — you may not have the latest version of QuadGK
INFO: Use `Pkg.update()` to get the latest versions of your packages
INFO: Package GR is already installed
INFO: METADATA is out-of-date — you may not have the latest version of GR
INFO: Use `Pkg.update()` to get the latest versions of your packages


In [3]:
using Plots
gr()

In [4]:
plot([
        x->real(bw(x,0.892^2,0.15)),
        x->imag(bw(x,0.892^2,0.15))
    ],
    linspace(0.4,1.3,100),
    lab = ["real", "imag"]
)

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 0.4 
 
 
 0.6 
 
 
 0.8 
 
 
 1.0 
 
 
 1.2 
 
 
 -2 
 
 
 0 
 
 
 2 
 
 
 4 
 
 
 6 
 
<polyline clip-path="url(#clip2102)" style="stroke:#009af9; stroke-width:1; stroke-opacity:1; fill:none" points="
 29.2591,180.71 34.8254,179.343 40.3916,177.934 45.9578,176.482 51.524,174.984 57.0902,173.441 62.6565,171.852 68.2227,170.214 73.7889,168.529 79.3551,166.796 
 84.9213,165.015 90.4876,163.187 96.0538,161.313 101.62,159.396 107.186,157.438 112.752,155.444 118.319,153.419 123.885,151.372 129.451,149.312 135.017,147.252 
 140.584,145.207 146.15,143.199 151.716,141.252 157.282,139.396 162.848,137.671 168.415,136.122 173.981,134.805 179.547,133.789 185.113,133.153 190.68,132.992 
 196.246,133.415 201.812,134.547 207.378,136.528 212.944,139.508 218.511,143.644 224.077,149.091 229.643,155.988 235.209,164.445 240.776,174.523 246.342,186.215 
 251.908,199.425 257.474,213.963 263.04,229.541 268.607,245.786 274.173,262.267 279.739,278.533 285.305,294.15 290.872,308.743 296.438,322.02 302.004,333.784 
 307.57,343.937 313.136,352.468 318.703,359.435 324.269,364.946 329.835,369.139 335.401,372.168 340.967,374.192 346.534,375.36 352.1,375.813 357.666,375.677 
 363.232,375.06 368.799,374.06 374.365,372.755 379.931,371.216 385.497,369.497 391.063,367.647 396.63,365.703 402.196,363.697 407.762,361.654 413.328,359.594 
 418.895,357.534 424.461,355.485 430.027,353.46 435.593,351.464 441.159,349.504 446.726,347.585 452.292,345.709 457.858,343.879 463.424,342.096 468.991,340.36 
 474.557,338.673 480.123,337.034 485.689,335.442 491.255,333.897 496.822,332.397 502.388,330.943 507.954,329.532 513.52,328.163 519.087,326.836 524.653,325.549 
 530.219,324.3 535.785,323.089 541.351,321.914 546.918,320.773 552.484,319.667 558.05,318.592 563.616,317.549 569.183,316.536 574.749,315.553 580.315,314.597 
 
 "/>
<polyline clip-path="url(#clip2102)" style="stroke:#e26f46; stroke-width:1; stroke-opacity:1; fill:none" points="
 29.2591,229.484 34.8254,228.425 40.3916,227.3 45.9578,226.103 51.524,224.83 57.0902,223.473 62.6565,222.025 68.2227,220.478 73.7889,218.825 79.3551,217.055 
 84.9213,215.159 90.4876,213.125 96.0538,210.941 101.62,208.592 107.186,206.065 112.752,203.341 118.319,200.403 123.885,197.231 129.451,193.803 135.017,190.095 
 140.584,186.082 146.15,181.736 151.716,177.028 157.282,171.927 162.848,166.402 168.415,160.42 173.981,153.951 179.547,146.965 185.113,139.441 190.68,131.362 
 196.246,122.724 201.812,113.539 207.378,103.845 212.944,93.7051 218.511,83.2224 224.077,72.5429 229.643,61.862 235.209,51.4268 240.776,41.5326 246.342,32.5122 
 251.908,24.7169 257.474,18.4891 263.04,14.1292 268.607,11.8626 274.173,11.811 279.739,13.9774 285.305,18.245 290.872,24.3929 296.438,32.1234 302.004,41.0951 
 307.57,50.9566 313.136,61.3737 318.703,72.049 324.269,82.7333 329.835,93.2287 335.401,103.387 340.967,113.103 346.534,122.312 352.1,130.976 357.666,139.081 
 363.232,146.631 368.799,153.64 374.365,160.133 379.931,166.136 385.497,171.682 391.063,176.802 396.63,181.527 402.196,185.889 407.762,189.917 413.328,193.638 
 418.895,197.079 424.461,200.262 430.027,203.21 435.593,205.944 441.159,208.48 446.726,210.837 452.292,213.028 457.858,215.069 463.424,216.971 468.991,218.746 
 474.557,220.404 480.123,221.956 485.689,223.408 491.255,224.769 496.822,226.047 502.388,227.246 507.954,228.375 513.52,229.437 519.087,230.437 524.653,231.381 
 530.219,232.272 535.785,233.113 541.351,233.909 546.918,234.663 552.484,235.377 558.05,236.053 563.616,236.696 569.183,237.306 574.749,237.885 580.315,238.437 
 
 "/>
 
 
 
 
 real 
 
 
 
 imag

## Integral over Dalitz plot

In [5]:
using QuadGK

In [6]:
# first, lambda
λ(x,y,z)=x^2+y^2+z^2-2*x*y-2*y*z-2*z*x
# second, integrate phase space 
function projectDalitzX(ampl,s1,s,m1sq,m2sq,m3sq)
    s1 < (sqrt(m2sq)+sqrt(m3sq))^2 && return 0
    s1 > (sqrt(s)-sqrt(m1sq))^2 && return 0
    ranges = m1sq+m2sq+(s-s1-m1sq)*(s1+m2sq-m3sq)/(2*s1)-sqrt(λ(s,s1,m1sq)*λ(s1,m2sq,m3sq))/(2*s1)*[-1.,1.]
    return quadgk(
        s3->abs(ampl(s1,s3,s,m1sq,m2sq,m3sq))^2,
        ranges[2],
        ranges[1])[1]/(8*π)^2/s;
end

projectDalitzX (generic function with 1 method)

In [232]:
# masses
mB = 5.279;
mJpsi = 3.01;
mπ = 0.139;
mK = 0.485;
mKs = 0.893;
ΓKs = 0.05;
mμ = 0.105;
# projection of the dalitz plot for the B to J/psi pi K
bw_amp(s1,s3,s,m1sq,m2sq,m3sq) = bw(s1,mKs^2,ΓKs)
phsp_s1(s1,s,m1sq,m2sq,m3sq) = sqrt(λ(s1,m2sq,m3sq)*λ(s,s1,m1sq))/s1/(8π)^2/s;
plot([
        e1->log(projectDalitzX(bw_amp,e1^2,mB^2,mJpsi^2,mπ^2,mK^2)),
        e1->log(abs(bw(e1^2,mKs^2,ΓKs))^2*phsp_s1(e1^2,mB^2,mJpsi^2,mπ^2,mK^2))
    ],
    linspace((mπ+mK)+0.01,(mB-mJpsi)-0.01,200))

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 0.9 
 
 
 1.2 
 
 
 1.5 
 
 
 1.8 
 
 
 2.1 
 
 
 -10.0 
 
 
 -7.5 
 
 
 -5.0 
 
 
 -2.5 
 
<polyline clip-path="url(#clip7702)" style="stroke:#009af9; stroke-width:1; stroke-opacity:1; fill:none" points="
 45.7332,202.817 48.4196,191.46 51.1059,183.744 53.7922,177.62 56.4786,172.367 59.1649,167.646 61.8513,163.267 64.5376,159.111 67.2239,155.099 69.9103,151.17 
 72.5966,147.278 75.283,143.387 77.9693,139.463 80.6556,135.476 83.342,131.397 86.0283,127.198 88.7147,122.848 91.401,118.317 94.0873,113.568 96.7737,108.564 
 99.46,103.258 102.146,97.5979 104.833,91.523 107.519,84.961 110.205,77.828 112.892,70.0294 115.578,61.4678 118.264,52.07 120.951,41.8617 123.637,31.1592 
 126.323,20.9809 129.01,13.5615 131.696,11.811 134.382,16.6302 137.069,25.7699 139.755,36.4268 142.441,46.9899 145.128,56.8431 147.814,65.8427 150.5,74.0228 
 153.187,81.4712 155.873,88.2831 158.56,94.5456 161.246,100.334 163.932,105.711 166.619,110.73 169.305,115.436 171.991,119.864 174.678,124.048 177.364,128.013 
 180.05,131.781 182.737,135.373 185.423,138.806 188.109,142.093 190.796,145.247 193.482,148.28 196.168,151.202 198.855,154.022 201.541,156.746 204.227,159.382 
 206.914,161.937 209.6,164.416 212.286,166.823 214.973,169.165 217.659,171.444 220.345,173.664 223.032,175.829 225.718,177.943 228.404,180.007 231.091,182.026 
 233.777,184 236.463,185.933 239.15,187.827 241.836,189.683 244.522,191.504 247.209,193.291 249.895,195.045 252.581,196.769 255.268,198.463 257.954,200.129 
 260.64,201.768 263.327,203.381 266.013,204.97 268.699,206.535 271.386,208.077 274.072,209.598 276.758,211.098 279.445,212.577 282.131,214.037 284.818,215.479 
 287.504,216.902 290.19,218.308 292.877,219.698 295.563,221.072 298.249,222.43 300.936,223.773 303.622,225.102 306.308,226.417 308.995,227.718 311.681,229.007 
 314.367,230.283 317.054,231.547 319.74,232.8 322.426,234.041 325.113,235.271 327.799,236.491 330.485,237.701 333.172,238.901 335.858,240.091 338.544,241.272 
 341.231,242.445 343.917,243.609 346.603,244.765 349.29,245.912 351.976,247.053 354.662,248.185 357.349,249.311 360.035,250.43 362.721,251.542 365.408,252.648 
 368.094,253.748 370.78,254.842 373.467,255.931 376.153,257.014 378.839,258.092 381.526,259.166 384.212,260.234 386.898,261.298 389.585,262.358 392.271,263.414 
 394.957,264.467 397.644,265.515 400.33,266.561 403.016,267.603 405.703,268.643 408.389,269.68 411.076,270.715 413.762,271.747 416.448,272.778 419.135,273.807 
 421.821,274.834 424.507,275.861 427.194,276.886 429.88,277.911 432.566,278.935 435.253,279.96 437.939,280.984 440.625,282.009 443.312,283.034 445.998,284.061 
 448.684,285.089 451.371,286.119 454.057,287.15 456.743,288.184 459.43,289.221 462.116,290.261 464.802,291.304 467.489,292.352 470.175,293.404 472.861,294.461 
 475.548,295.523 478.234,296.591 480.92,297.666 483.607,298.748 486.293,299.838 488.979,300.936 491.666,302.044 494.352,303.162 497.038,304.292 499.725,305.433 
 502.411,306.587 505.097,307.756 507.784,308.941 510.47,310.142 513.156,311.363 515.843,312.604 518.529,313.867 521.215,315.155 523.902,316.47 526.588,317.815 
 529.274,319.193 531.961,320.609 534.647,322.065 537.334,323.569 540.02,325.124 542.706,326.739 545.393,328.422 548.079,330.183 550.765,332.036 553.452,333.996 
 556.138,336.084 558.824,338.327 561.511,340.762 564.197,343.438 566.883,346.429 569.57,349.847 572.256,353.875 574.942,358.848 577.629,365.478 580.315,375.813 
 
 "/>
<polyline clip-path="url(#clip7702)" style="stroke:#e26f46; stroke-width:1; stroke-opacity:1; fill:none" points="
 45.7332,202.817 48.4196,191.46 51.1059,183.744 53.7922,177.62 56.4786,172.367 59.1649,167.646 61.8513,163.267 64.5376,159.111 67.2239,155.099 69.9103,151.17 
 72.5966,147.278 75.283,143.387 77.9693,139.463 80.6556,135.476 83.342,131.397 86.0283,127.198 88.7147,122.848 91.401,118.317 94.0873,113.568 96.7737,108.564 
 99.46,103.258 102.146,97.5979 104.833,91.523 

## Phase space MonteCarlo

In [233]:
# prepare array of weights
nbinsX = 300;
ms3_bins = linspace((mJpsi+mπ)^2,(mB-mK)^2,nbinsX+1)
density(s3) = sqrt(λ(mB^2,s3,mK^2)*λ(s3,mJpsi^2,mπ^2))/s3
weights = [density((ms3_bins[i]+ms3_bins[i+1])/2) for i in 1:nbinsX];

In [234]:
using StatsBase

In [235]:
nEvents = 1000000;
inds = StatsBase.sample(1:nbinsX, Weights(weights), nEvents)
s3_data = [ms3_bins[i]+rand()*(ms3_bins[i+1]-ms3_bins[i]) for i in inds];

In [236]:
histogram(s3_data, bins=200)

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 12 
 
 
 15 
 
 
 18 
 
 
 21 
 
 
 0 
 
 
 2500 
 
 
 5000 
 
 
 7500 
 
 
 10000 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 y1

In [237]:
using DataFrames

In [238]:
### Kinematic function
λ(x,y,z) = x^2+y^2+z^2-2*x*y-2*y*z-2*z*x;
density(s1,s,m1sq,m2sq,m3sq) = sqrt(λ(s,s1,m1sq)*λ(s1,m2sq,m3sq))/s1
s3_ranges(s1,s,m1sq,m2sq,m3sq) = m1sq+m2sq+(s-s1-m1sq)*(s1+m2sq-m3sq)/(2*s1)-sqrt(λ(s,s1,m1sq)*λ(s1,m2sq,m3sq))/(2*s1)*[1.,-1.]

### Functions to boost and rotate
Rz(θ) = [cos(θ) -sin(θ) 0 ; sin(θ) cos(θ) 0 ; 0 0 1];
Ry(θ) = [cos(θ) 0 sin(θ); 0 1 0 ; -sin(θ) 0 cos(θ)];
function Λz(γ)
    β = sqrt(1-1/γ^2);
    [γ 0 0 β*γ; 0 1 0 0; 0 0 1 0; β*γ 0 0 γ]
end
Rot4(R3) = [1 [0 0 0]; [0; 0; 0] R3];
boost_rotate(γ,θ,ϕ) = Rot4(Rz(ϕ)*Ry(θ))*Λz(γ);

# scalar products
mprod(P1,P2) = P1[1]*P2[1]-P1[2:4]⋅P2[2:4]
mnorm(P1) = mprod(P1,P1);

In [239]:
nEvents = 100000;

# data storage
tree = DataFrame(
    [Array{Float64,1} for i in 1:6],
    [:P1, :P2, :P3, :P4, :Q1, :Q2],
    nEvents);

# s-data
# prepare array of weights
nbinsX = 300;
ms_bins = linspace((mJpsi+mπ)^2,(mB-mK)^2,nbinsX+1)
density(s) = sqrt(λ(mB^2,s,mK^2)*λ(s,mJpsi^2,mπ^2))/s
weights = [density((ms_bins[i]+ms_bins[i+1])/2) for i in 1:nbinsX];
# generate nEvents-s-values
inds = StatsBase.sample(1:nbinsX, Weights(weights), nEvents)
s_data = [ms3_bins[i]+rand()*(ms3_bins[i+1]-ms3_bins[i]) for i in inds];

# main look to fill trees with four-vectors
mπsq = mπ^2;
for i in 1:nEvents

    s = s_data[i];
    
    # P2-frame
    cosT4 = 2*rand()-1.; sinT4 = sqrt(1-cosT4^2); θ4 = acos(cosT4);
    phi4 = π*(2*rand()-1.);
    p4 = sqrt(λ(mB^2, s, mK^2))/(2*mB);
    e4 = sqrt(p4^2+mK^2);  e13 = sqrt(p4^2+s);
    tree[i,:P2] = [mB,0,0,0];
    tree[i,:P4] = [e4,-p4*sinT4*cos(phi4),-p4*sinT4*sin(phi4),-p4*cosT4];

    # from P1-P3 rest frame
    γ13 = e13/sqrt(s);
    cosT1 = 2*rand()-1.; sinT1 = sqrt(1-cosT1^2); θ1 = acos(cosT1);
    phi1 = π*(2*rand()-1.);
    p1 = sqrt(λ(s, mJpsi^2, mπ^2)/(4*s));
    e1 = sqrt(p1^2+mJpsi^2); e3 = sqrt(p1^2+mπ^2);
    tree[i,:P1] = boost_rotate(γ13,θ4,phi4)*[e1,p1*sinT1*cos(phi1),p1*sinT1*sin(phi1),p1*cosT1];
    tree[i,:P3] = boost_rotate(γ13,θ4,phi4)*[e3,-p1*sinT1*cos(phi1),-p1*sinT1*sin(phi1),-p1*cosT1];
    
    # from Q1-Q2 rest frame
    γ1 = e1/mJpsi;
    cosTμ = 2*rand()-1.; sinTμ = sqrt(1-cosTμ^2);
    phiμ = π*(2*rand()-1.);
    pμ = sqrt(λ(mJpsi^2, mμ^2, mμ^2))/(2*mJpsi);
    eμ = sqrt(pμ^2+mμ^2);
    tree[i,:Q1] = boost_rotate(γ13,θ4,phi4)*boost_rotate(γ1,θ1,phi1)*[eμ, pμ*sinTμ*cos(phiμ), pμ*sinTμ*sin(phiμ), pμ*cosTμ];
    tree[i,:Q2] = boost_rotate(γ13,θ4,phi4)*boost_rotate(γ1,θ1,phi1)*[eμ,-pμ*sinTμ*cos(phiμ),-pμ*sinTμ*sin(phiμ),-pμ*cosTμ];    
end

In [240]:
tree[1,:]

1×6 DataFrames.DataFrame
│ Row │ P1                                        │ P2                     │
├─────┼───────────────────────────────────────────┼────────────────────────┤
│ 1   │ [3.31022, -0.406927, 0.0615314, -1.31457] │ [5.279, 0.0, 0.0, 0.0] │

│ Row │ P3                                    │
├─────┼───────────────────────────────────────┤
│ 1   │ [1.24929, 0.47942, -0.50356, 1.02859] │

│ Row │ P4                                        │
├─────┼───────────────────────────────────────────┤
│ 1   │ [0.719484, -0.072493, 0.442029, 0.285986] │

│ Row │ Q1                                      │
├─────┼─────────────────────────────────────────┤
│ 1   │ [1.06105, 0.344754, 0.710567, 0.700741] │

│ Row │ Q2                                        │
├─────┼───────────────────────────────────────────┤
│ 1   │ [2.24918, -0.751681, -0.649035, -2.01532] │

In [241]:
# check if s-distribution is the same as input
histogram([
        [mnorm(tree[i,:P1]+tree[i,:P3]) for i=1:nEvents],
        s_data
        ],
     nbins=200,fillalpha=0.7)

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 12 
 
 
 15 
 
 
 18 
 
 
 21 
 
 
 0 
 
 
 250 
 
 
 500 
 
 
 750 
 
 
 1000 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 y1 
 
 
 
 
 y2

In [313]:
# check energy conservation
histogram([tree[i,:P2][1]-tree[i,:P3][1]-tree[i,:P4][1]-tree[i,:Q1][1]-tree[i,:Q2][1] for i=1:nEvents],
    nbins=200,
    xlims=(-0.1,0.1),lab="eB-sum_e")
# check B-invariant mass
histogram!([mB^2 - mnorm(tree[i,:P4]+tree[i,:P3]+tree[i,:Q1]+tree[i,:Q2]) for i=1:nEvents],
    nbins=200,
xlims=(-0.1,0.1),lab="mB^2-|sumP|^2")

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 -0.10 
 
 
 -0.05 
 
 
 0.00 
 
 
 0.05 
 
 
 0.10 
 
 
 0 
 
 
 10000 
 
 
 20000 
 
 
 30000 
 
 
 40000 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 eB-sum_e 
 
 
 
 
 mB^2-|sumP|^2

In [243]:
# check dalitz plot
histogram2d(
    [mnorm(tree[i,:P3]+tree[i,:P4]) for i in 1:nEvents],
    [mnorm(tree[i,:Q1]+tree[i,:Q2]+tree[i,:P3]) for i in 1:nEvents],
    bins=(100,100),
    xlabel="t [pi Jpsi] (GeV)",
    ylabel="s [pi K] (GeV)",
    title = "Jpsi pi K phase space Dalitz plot"
    )

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 1 
 
 
 2 
 
 
 3 
 
 
 4 
 
 
 5 
 
 
 12 
 
 
 15 
 
 
 18 
 
 
 21 
 
 
 Jpsi pi K phase space Dalitz plot 
 
 
 t [pi Jpsi] (GeV) 
 
 
 s [pi K] (GeV) 
 
 
<image width="476" height="329" xlink:href="data:;base64,
iVBORw0KGgoAAAANSUhEUgAAAdwAAAFJCAYAAAAxCJwFAAAgAElEQVR4nO29eXiV9b3u/ShIJjSQ
QEBMCGEKUwMEkcHIpFCEIsM2VopusWq1esqunnZvj7u7PXX3aG19rdW3tlSruJWthW4GEUSiTEYG
gQApU8IYEhFCEgiQhCDQ88f711r3p1efV92PWu/Pf9zXWlnrGX881/pc9/eSwHxuJLbq/Jf4bGzC
JHnd1KyzkjWebylZRUOrmH/f2muvvKbqRJpkI4ZvkOzpJRMkK+h4TLLK01dI9vaRZMl6pcqmBiV1
mjVcOC/ZwUsPa3bmrUskNMaYLzCXft5fwBhjjPkq4AXXGGOMiQAvuMYYY0wEeME1xhhjIsDiyaek
xaVtxPzpnTROXvevXZMkm1L4hmQ1u3Mkq6jMlKxHz3363o86xPy7tKqzvKbpgspWY/tvlax1+knJ
juzPlqzrsG2SnW/QbX1z2VjJemaohPX6nh6Sjcs8KlkZyF+5besk+9GfW0uWErSSbM3ZP0p24eJJ
Xx/GmM8MP+EaY4wxEeAF1xhjjIkAL7jGGGNMBHjBNcYYYyLAUshfpaXIUEVDpsqrahq1VYmYNnO+
ZCd3dpFsw7YBkiVf9rFkOVd+KFl9XOtTn5Eb5TW71gyRLLFVs2QnTl8uGYlap+vaSNbcnBDq75XX
tZcsPbFJsrKT+hnDO1VJtu6IymXtEnXbCie9JdnLC7Xha/KgTZIVbR8o2exDFyU7dEmZZMca1vl6
M+YrjJ9wjTHGmAjwgmuMMcZEgBdcY4wxJgK84BpjjDERoLVDX0H6pdwigtSv+unoOBKk2iU3SkYj
6yrfUxkqXnIKgiBo/PgyydKTz0i2try3ZBOHro/5N7VWEQkJKhYtLu0n2czLT0t29pwKUu8e7CrZ
rno91f59XLFkp06mSkbb3/Sxfm5WSoNkeZk62i9xnDZhjS4tl2zxlsGSHW/W4/P+a9pSteGXwyRr
e/loOal++cHX5HW7m3Q/b2x8JS45bwHLmC8ZfsI1xhhjIsALrjHGGBMBXnCNMcaYCPCCa4wxxkTA
V068+H973iviCo11I0hoIkjyaQvC0ZG6dpKNunOJZK88MyPU58YzJPuAZKv295Tsjqn6mecbEiU7
sFvfu75SR/YRWa11+/Nz90hG0lTujA8km//EzZIV9PuzZHsOdJNs63FtuJqYu1syksm63LlXst/9
cJpkM8YVSfb8svGSTcvT8YYLSlWwG9j+eMy/aRsaLuj/n39a8TvJLFwZ8/ngJ1xjjDEmArzgGmOM
MRHgBdcYY4yJAC+4xhhjTAT8XcsTky5/QASp/7x/obyu8biOf6v5qINkJDklXaZiTW1ja8lGDN8g
2fI110nWdEEbmeIbpIIgCF5ePUqy63NiJanyat2GrNQTknVsVyNZQtJZyaqrMySjEYDPzpkuGUlJ
JHDdc298o1IQrJ03UTIaT7h8l7Y2Jbc8L1lSC81ov/S/drNkh0tzJVtapq1f1HpFIxo3LbxBMqKy
vm3Mv2lk45ixqyW7e/Y/SNYuQS/7V0+tlKy+SaU2C1fGfHL8hGuMMcZEgBdcY4wxJgK84BpjjDER
4AXXGGOMiYC/GwFiZvosEaTaJ+qIvck5FZKR5FR7Vke4kWxD1DZrS1NhwXuS7S3vLhkJTAc/ukoy
GhW459TlMf/OT9cGLWrLIlFrRE+VnI7WqDQWL/P8NUgYWjCnMNTrfv2bmZLRuL+vd6qXrLham6vG
XaX7mPZLSZ2OT2yfcEGy+BaoIAiCTmn6GSdOXy4ZNZBd0Ua3o7kp9pz61fqr5TW9U1XgG91Nxw5S
gxb9vQeHqTRG5+LNpXquNF84JdnZc4f/bu43xnwS/IRrjDHGRIAXXGOMMSYCvOAaY4wxEeAF1xhj
jImAL6XEMCZZR+y9PHmTvK5o+0DJSAaqP61yzBv7VGgiCYkaf9JgFN27B7tKltvmpGRE2HF/8c1N
kwfpPikp6/WJ/36nbiqcnWtUQWx+sTZokbzTOa9MspVFoyQrqU2TLDulSbKeaSovhRXi0hP17xFD
B+g4vaqKzpJtB7lo6phVktH+I557d0zMv2n7if7QyFV1Qvdn5RkVuhrPt5CMGsPoOFLDVeOFi5JV
X2iUbGPjS1/K+5Ixfws/4RpjjDER4AXXGGOMiQAvuMYYY0wEeME1xhhjIuALLye0uLSNCFK/7v5N
eV1uWxWaaIRbc3OCZPN29ZWMBKn8XB1XVrxDR8IV/myZZPN/NEGyt49oCxK1JY0fqS1ViRk6Tq54
aeyoN9r+9Myjkh3YrWPy6kCiIXp13S/ZaxuHSja8U5VkJGaVVqmARO1T+94ZIlnY8Ynbqq+ULAtE
N2qfIkic697hI8nOntNzj2S6480qK80a8X7Mv8PKb30H7JCstqqjZCQOEv1nbZHsdJGeKy1TdLwj
jTak9q3bt+n+zLio39lylfmy4SdcY4wxJgK84BpjjDER4AXXGGOMiQAvuMYYY0wEfOGlg2+2+Z5I
U1OydPRXu2RtrCmY+E6oz/jx7Nv0vR1qJcsEMYuglqEh2Qcko+amhStHS5Z1uW4vyUoVlZkx/+7R
c5+85uXVo/R9DTrq7meFb0o2d8VY/W4gG1Fr0fFmFZBu6bNTMpJontieJdnPC3ZJRpBIVQZNS8kt
dfQitTT1e1zH7j1/2yDJBmSoNNX0sUpTJGaRcNXuymMx/6Z2q/LqDvqZMHqRWrUWVuh+6pOq+2Ra
njZtkXCVkVEt2Z4D3SQjqY1ISjgnWSpcF99ZoRLjzkv+LNmxhnVf+Huf+fvDT7jGGGNMBHjBNcYY
YyLAC64xxhgTAV5wjTHGmAhQo+JzJKf1jSJIDUq7IK+rbdaRZj0zjknWMkPlkNUvTZLsB19/V7Kl
G4ZJNubqUsmefH2aZAUd9btQy9DmTSrbJLVQUYXG/VGb08D2sePpnnz7ennNrb32SkatUk8v0Was
sCPhaBQfyUu592kL0vZndJ/kp6nfQnIVNVcRFQ2t4DO0GYkkpNTHVNQhcWzdkUzJSDg6CILdLz9Q
8Wf2Ixtj/k3SVF7mYcmIpWW9JfvRdZslo1a2sI1UrdN19OSildrwNb3HIcmys7SVLF4IDIIgePfg
AMl+Py6c1PXstllyr6k9p/eaJad/Y7nKfGb4CdcYY4yJAC+4xhhjTAR4wTXGGGMiwAuuMcYYEwFf
KCHggQ4qMoy7Spt9qPEpM1uFkXNNKlcVbR8oWeGktyQ7eVjHgZG4UdvYWrIRwzdItnNbv1DvDdvS
dP/1KyWr+ShW8iGxiGQeYle9+nQvvPiaZEt/qu1T42cskmztvImSdUrTY0vj+Wj83ZixqyVbWTQq
1HsXHVJ55+mf/l6yo2v6SEbnz9j+WyWjc4Xakqqg9YrGQJ46GTvKcd8x3QaCWqtW7ddxjCSShW1b
W1+ZLRmNy6RWLfp71NRG7Vh9Ox+SbOfhLpLRNdV4XkcgfrdQz1s6p17Y21ayJadnSxYE579Q91fz
+eMnXGOMMSYCvOAaY4wxEeAF1xhjjIkAL7jGGGNMBHxuP+ontuosgtRz3VXA6Zl2XLLKepUWiEWV
2jBzXYaKK+P76PiuhCRtHuo4UkfC1X3QVbLSnX0lo7aop

In [244]:
using GSL

In [245]:
# Clebsches and d-function
function ClebschGordon(j1,m1,j2,m2,j,m)
    factor = (j1+j2-m)%2==1 ? -1 : 1
    factor*sqrt(2*j+1)*sf_coupling_3j(2*j1,2*j2,2*j,2*m1,2*m2,-2*m)
end

function Wignerdhat(j,m1,m2,cosθ)
#    print("WignerD(",j,",",m1,",",m2,",",cosθ,")")
    (m1 < 0) && print("Error: Something is wrong with Wignerdhat!")
    (m2 == 0) && sqrt(4*π/(2*j+1))*sf_legendre_sphPlm(j,m1,cosθ)/sqrt(1-cosθ*cosθ)^(m1)
end


Wignerdhat (generic function with 1 method)

In [246]:
# tensor pre-factors
function Cs(p1,p2,p3,p4)
    s = mnorm(p3+p4)
    p3-p4 - (mnorm(p3)-mnorm(p4))/s*(p3+p4)
end
Bs(p1,p2,p3,p4) = p3+p4
function Ds(p1,p2,p3,p4)
    # minus is removed due to the crossing
    # i is removed to make amplitude real
    [-sum((σ-ν)*(σ-ρ)*(σ-μ)*(ν-ρ)*(ν-μ)*(ρ-μ)*
            p1[σ]*p2[ν]*p3[ρ]
        for σ=1:4, ν=1:4, ρ=1:4)
        for μ=1:4]
end

# kinematical functions
function ζs(j,l,s,zs)
    (l < j-1 || l > j+1) && return 0;
    # three cases
    (l==j) && return 0;
    ps = sqrt(λ(s,mJpsi^2,mB^2)/(4*s))
    qs = sqrt(λ(s,mπ^2,mK^2)/(4*s))
    (l==j-1) && return (ps*qs)^(j-1)/(4*π)*sqrt((2*j+1)*(2*j-1)/2.0)*
        ClebschGordon(j-1,0,1,1,j,1)*Wignerdhat(j,1,0,zs)
    (l==j+1) && return λ(s,mJpsi^2,mB^2)*(ps*qs)^(j-1)/(4*π)*sqrt((2*j+1)*(2*j+3)/2.0)*
        ClebschGordon(j+1,0,1,1,j,1)*Wignerdhat(j,1,0,zs)
end
function βs(j,l,s,zs)
    (l < j-1 || l > j+1) && return 0;
    # three cases
    (l == j) && return 0;
    ps = sqrt(λ(s,mJpsi^2,mB^2)/(4*s))
    qs = sqrt(λ(s,mπ^2,mK^2)/(4*s))
    (l == j-1) && return (ps*qs)^(j)/λ(s,mJpsi^2,mB^2)*
        (s+mJpsi^2-mB^2)/sqrt(2)*
        4/(4*π)*sqrt((2*j+1)*(2*j-1)/2.0)*(
            ClebschGordon(j-1,0,1,0,j,0)*Wignerdhat(j,0,0,zs)/sqrt(2)+
            ClebschGordon(j-1,0,1,1,j,1)*Wignerdhat(j,1,0,zs)*zs
        )
    (l == j+1) && return (ps*qs)^(j)*
        (s+mJpsi^2-mB^2)/sqrt(2)*
        4/(4*π)*sqrt((2*j+1)*(2*j+3)/2.0)*(
            ClebschGordon(j+1,0,1,0,j,0)*Wignerdhat(j,0,0,zs)/sqrt(2)+
            ClebschGordon(j+1,0,1,1,j,1)*Wignerdhat(j,1,0,zs)*zs
        )
end
function δs(j,l,s,zs)
    (l < j-1 || l > j+1) && return 0;
    # three cases
    (l == j-1 || j == j+1) && return 0;
    ps = sqrt(λ(s,mJpsi^2,mB^2)/(4*s))
    qs = sqrt(λ(s,mπ^2,mK^2)/(4*s))
    (l == j) && return (ps*qs)^(j-1)/(4*π)*sqrt((2*j+1)*(2*j-1)/2.0)*
        ClebschGordon(j-1,0,1,1,j,1)*Wignerdhat(j,1,0,zs)
    (l == j+1) && return (ps*qs)^(j-1)/(4*π)*sqrt(2.0)*(2*j+1)*
        Wignerdhat(j,1,0,zs)
end

function Zjl(j,l,p1,p2,p3,p4)
    s = mnorm(p3+p4)
    t = mnorm(p1+p3)
    u = mB^2+mJpsi^2+mπ^2+mK^2-s-t
    zs = (s*(t-u)+(mπ^2-mK^2)*(mJpsi^2-mB^2))/sqrt(λ(s,mB^2,mJpsi^2)*λ(s,mπ^2,mK^2))
#    print("zs = ",zs,"\n")
    Cs(p1,p2,p3,p4)*ζs(j,l,s,zs)+
        Bs(p1,p2,p3,p4)*βs(j,l,s,zs)+
        Ds(p1,p2,p3,p4)*δs(j,l,s,zs)
end

Zjl (generic function with 1 method)

In [272]:
# definition of the amplitude and square of matrix element
# i.e. contracted with lepton current
function ampl(p1,p2,p3,p4,pars)
    amp = [0,0,0,0]
    for j in 1:length(pars)
        ps = pars[j]
        (length(ps) != 3) && continue
        amp += (ps[1] == 0) ? 0 : ps[1]*Zjl(j,j-1,p1,p2,p3,p4);
        amp += (ps[2] == 0) ? 0 : ps[2]*Zjl(j,j,  p1,p2,p3,p4);
        amp += (ps[3] == 0) ? 0 : ps[3]*Zjl(j,j+1,p1,p2,p3,p4);
    end
    amp
end

function amplsq(p1,p2,p3,p4,q1,q2,pars)    
    amp = ampl(p1,p2,p3,p4,pars)
    ampc = conj(amp);
    real(mprod(amp,q1)*mprod(ampc,q2)+mprod(amp,q2)*mprod(ampc,q1)-mprod(amp,ampc)*mJpsi^2/2)
#    real(amp[2:4]⋅ampc[2:4])
end

amplsq (generic function with 1 method)

In [273]:
Zjl(5,6,tree[2,:P1],tree[2,:P2],tree[2,:P3],tree[2,:P4])

4-element Array{Float64,1}:
  3573.89 
 -1924.61 
  -597.485
 -1998.09 

In [274]:
amplsq(tree[1,:P1],tree[1,:P2],
       tree[1,:P3],tree[1,:P4],
       tree[1,:Q1],tree[1,:Q2],
       [[0,0,1]])

124.04546144302651

In [275]:
hists = histogram(
    [mnorm(tree[i,:P3]+tree[i,:P4]) for i in 1:nEvents],
    weights = [amplsq(tree[i,:P1],tree[i,:P2],
    tree[i,:P3],tree[i,:P4],tree[i,:Q1],tree[i,:Q2],
            [[1,0,0]]) for i=1:nEvents],
    nbins=200,norm=true, lab="j=1, l=0"
    )
histogram!(
    [mnorm(tree[i,:P3]+tree[i,:P4]) for i in 1:nEvents],
    weights = [amplsq(tree[i,:P1],tree[i,:P2],
            tree[i,:P3],tree[i,:P4],tree[i,:Q1],tree[i,:Q2],
            [[0,0,1]]) for i=1:nEvents],
nbins=200,norm=true, lab="j=1, l=2",fillalpha=0.4
    )

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 1 
 
 
 2 
 
 
 3 
 
 
 4 
 
 
 5 
 
 
 0.0 
 
 
 0.1 
 
 
 0.2 
 
 
 0.3 
 
 
 0.4 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 j=1, l=0 
 
 
 
 
 j=1, l=2

In [267]:
histogram2d(
    [mnorm(tree[i,:P3]+tree[i,:P4]) for i in 1:nEvents],
    [mnorm(tree[i,:P1]+tree[i,:P3]) for i in 1:nEvents],
    weights = [amplsq(tree[i,:P1],tree[i,:P2],
            tree[i,:P3],tree[i,:P4],tree[i,:Q1],tree[i,:Q2],
    [[0,0,1]]) for i=1:nEvents],
#    weights = [abs(bw(mnorm(tree[i,:P3]+tree[i,:P4]),1.4^2,0.3))^2 for i=1:nEvents],
    nbins=50
    )

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 1 
 
 
 2 
 
 
 3 
 
 
 4 
 
 
 5 
 
 
 12 
 
 
 15 
 
 
 18 
 
 
 21 
 
 
<image width="491" height="364" xlink:href="data:;base64,
iVBORw0KGgoAAAANSUhEUgAAAesAAAFsCAYAAADsRpHBAAAXeElEQVR4nO3dS4xsiX0W8HOqq7v6
3bf73rl37p2nx8PEmDjYRo4DWbDIwskSsUqQIpQFIlJgwwoRFllFrGDPFgkJFpHYRIBQgDjEAU1s
jYcMmfHM3HncV99Hd9/uqq7nOWzIDqHvymPmb/v3W386VXXqVH1Vm/O1DT+0v779G32S+/WXhtHx
fuPv/asot/Zidrx+PI9y01/+1Si3+e/+dXa8D65GuT/99s9HufdPj6LcHx3vZMebTKPc2+33otzJ
5Z9HuVV31kZBgP9j8Hk/AQDg/01ZA0BxyhoAilPWAFCcsgaA4pQ1ABSnrAGgOGUNAMUpawAoTlkD
QHHKGgCKU9YAUJyyBoDifirXf17d/Va0kvX7fzNbZXr9H92JcqujG1Fu+P3vRrnZe9m61Oj1iyi3
vJuteB1/92ei3P0H16Nc9GY0TfOFL96Ocu+8kz2/K7vnUe6Dh9n7tj/Krpf/dOdWlPsfj5dR7v3B
R1Hu7vztKDedf/xT+b0AlflnDQDFKWsAKE5ZA0BxyhoAilPWAFCcsgaA4pQ1ABSnrAGgOGUNAMUp
awAoTlkDQHHKGgCK+4m6Yf9f2/71aBPiO//830THa28dRLnHv3ctyh194/0od/+/fjnKnZ1ciXJt
m01ldH12OVy/9SDKffTBK1Fuf3cc5dY35lHuD975K1FutlqLcs9vZ8/v+HI7yl3fmkS5x9OtKPfB
xWaUm3fZ+/vHp9nwSxdOsHx/+Z+j3OUiG8Tp+/lP1PcWJPyzBoDilDUAFKesAaA4ZQ0AxSlrAChO
WQNAccoaAIpT1gBQnLIGgOKUNQAUp6wBoDhlDQDFKWsAKO7HYr3maPtr0bzPvd95Nzre2pf2otx3
/tk3o9zXf+mPotz4XrbO9fb3s9WtX/iVP4hyD//nF6PcvQc3olyfjS01w7VVlJvMstWodx4/F+Ve
OziJcu+dHkW5abjOtb++iHLni2GUezJfj3Kr8P14Mss+7oPwWyF82GbRZbm7l8so92b/vSh3vrwf
5S7n6dpXeALhR8A/awAoTlkDQHHKGgCKU9YAUJyyBoDilDUAFKesAaA4ZQ0AxSlrAChOWQNAccoa
AIpT1gBQ3Od6Y/q23Yy2AN761t+Ijre/fx7ltg+yXBcOOCxnG1GuHWSLBv/trZ+Lctd2LqLcg4v9
KJcOVpyFAxNfvf4gyg0H2eDHn9x7IcpdG82i3O3xdpR7bW8c5R5cbkW5h9NsyGMZLmU8N8rO33H4
uO+eZ9fptVH2W3+aPb3mZJENeaRDMg/6p1HudPA4yl30T6LcyfT9KLdanUa5vlkaEME/awCoTlkD
QHHKGgCKU9YAUJyyBoDilDUAFKesAaA4ZQ0AxSlrAChOWQNAccoaAIpT1gBQnLIGgOI+1zWXXzv8
h9F+zm995cPoeC+/+GmU++ROtt60s3kZ5SbTbG3p3UfPRbmNtWz16PF0FOVuj7Pc3379gyj39vHN
KPfxeDPKXRstotwnk+x1XGTjTc3PHmTrXCfzbK1qOMjmoM4X2bpZahyOMp3Ms+Olq1Zt+O0xW2UH
fDTProNVkx3vrJlEub7JPm/jQXa8yyZb+7pYPYpy5/M7UW7VZY/b93MrXj+G/LMGgOKUNQAUp6wB
oDhlDQDFKWsAKE5ZA0BxyhoAilPWAFCcsgaA4pQ1ABSnrAGgOGUNAMV9rjd0/72v/Z3ojvw39rIb
1N8/P4hyh5vZDfm/fS8brPjK4WmUu32+F+XmXfYbatlnb9+ru+Mol5oss2GL6SobrEhf73iZ5c7m
2eOO1rJBiJd2plHug/NsuGSyyt6303B4Yzd7O5rd9Sz3/nk2bLEeLnmM1rLcZJm9H0+X2VLLrMly
F032/o6a7ATeHWSDQunX76Q/iXLT1XmUmy2z76vFKnvcvp8aBvn/wD9rAChOWQNAccoaAIpT1gBQ
nLIGgOKUNQAUp6wBoDhlDQDFKWsAKE5ZA0BxyhoAilPWAFCcsgaA4sK9nmfz6u4vR/M5VzYvo+N9
+85LUe4Lu9nqzJ88eD7KnYbrTafzUZT70ycbUW4t/Am1H64ofTzej3I3t7K1pTuT7Ane2MpWlM4W
2WjPTriSla5pZa+2ad4524pyRxvZEc/D17udXX7NlY3s9a7ClbZhuKa1t57lzsM1ra7Jcocb2dfW
SbhaNg7Xr9owd7W7kT1um63hzdvse3IQ5rbXr0W5yza7AJfdRfTGdV22dtj3cyte/xf+WQNAccoa
AIpT1gBQnLIGgOKUNQAUp6wBoDhlDQDFKWsAKE5ZA0BxyhoAilPWAFCcsgaA4n4kN0z/Jy/9g+jG
7vvrq+h4O8Msd2eSDWXcvcxe9lF2uOZimeWe28yGCu6Hz+8yOy3NbjjXcjTKnt9x+PzSoYzr4XlJ
L9ZBGLwzyYLXN7PjPV1kuXAno9kMf0pPw+vgbJGd551h9gQ/mWQveC18wc9vZhfqJ5fZQsekyXLp
QMeqz67op4NsUGjRZs9v1WRfMItmGuUm3WmUW/bZ8RarbJBk2WXHW3UXUa7vZlmuWf5EDIP4Zw0A
xSlrAChOWQNAccoaAIpT1gBQnLIGgOKUNQAUp6wBoDhlDQDFKWsAKE5ZA0BxyhoAilPWAFBcuMf0
bE7n2cjJ+iD7rfDO2Wf7ND8NV3sez9ai3Eb4Osbh+Mt2+HJH4U+t03m2trQezlUdjrLHHYdrZKnJ
Knt+F+H6VbpGdpyNBTXhWFUzDefI5uH7+2SezW4dbWTXc7rmdriencB5n11/T+bZiVlvsxOz22cX
aroKdqd9EuX2+70od9Jm61frfTb/14avY9Fma1WDJrteukF2wbRtdry+z47XD8Ln10+jC7Dvw+L6
nPhnDQDFKWsAKE5ZA0BxyhoAilPWAFCcsgaA4pQ1ABSnrAGgOGUNAMUpawAoTlkDQHHKGgCKe6aF
jLXBXnRD9Fk4VPBfHmaLC4smu7H7KHw5m+EN4PsmGyBIc5erLDcJhxTWmuy+8/sbWe44u999czTK
jhfuNzT3L7PjjbK3rXkwyxZE9ofZAV/Pdhni15EOiIS7Fs1e+DrCt6N5usguwK217Lf+ssseeX89
ex2zLjvP01V2AtPP76vt1Sj30Sob/NjsN6Pcssmu53R4Y7c9inKTcGhk2WfDIE0bLgCFn4/lapIF
w+ulCz8ifR9+MJtwuSnknzUAFKesAaA4ZQ0AxSlrAChOWQNAccoaAIpT1gBQnLIGgOKUNQAUp6wB
oDhlDQDFKWsAKE5ZA0Bxz7S6tTt6JcodT7NVknQ1atRuRLk/b29Hude6l6Pczlp2eu4sxlFuvc1+
G90YbkW5QbjpchrON+2vZ89vmo0ANSfhetOVcG3pbJ6tI/3MXva+PQ7Hgk7mWe76Zvb8xqtwnSt8
f0/C15Gulo3CC2sVzqrthe9v6mKZXVdd+PzW2uz1Puyyz3kbfq8N++w6HQ/Oo1y6HtaFK4bDJlvJ
asPvtfS8tOF/yOEg+57s+nS1LPue7JrsC6FP35Bwncs/awAoTlkDQHHKGgCKU9YAUJyyBoDilDUA
FKesAaA4ZQ0AxSlrAChOWQNAccoaAIpT1gBQ3DMNeeysXYtyx6tJlDttT6PcrL2Mcodd9vxmTXZj
9zvdwyi30+5Eud0mGyQ5XS6i3JVhdrx0oGOR3ce+GS+z4PVwOWIY/

## Analytical approach
Now, I get the same distributions analytically
 - Matrix element square is lorethz invariant
 - in s-channel CMS, amplitude is straightforward
 - integrals over phase space are analytical as long as amplitude depend on (s, sz)
 - avaraging over J/psi polarization is a little bit tricky (see comment in the cell below)

In [279]:
function amplsp_analytic(s,zs,pars)
    # states along z axis
    p = sqrt(λ(s,mB^2,mJpsi^2)/(4*s))
    E1 = sqrt(p^2+mJpsi^2)
    E2 = sqrt(p^2+mB^2)
    p1 = [E1, 0,0,p]
    p2 = [E2, 0,0,p]
    # scattered states
    q = sqrt(λ(s,mπ^2,mK^2)/(4*s))
    E3 = sqrt(q^2+mπ^2)
    E4 = sqrt(q^2+mK^2)
    sinθ = sqrt(1-zs^2)
    p3 = [E3,-q*sinθ,0,-q*zs]
    p4 = [E4, q*sinθ,0,q*zs]
    # amplitude
    amp = ampl(p1,p2,p3,p4,pars)
    ampc = conj(amp)
    # avaraging over muon angles is easy in J/psi rest frame
    # the answer there is \vec A \cdot \vec A (some numerical factor)
    # therefore, I have to find \vec A^2 in Jpsi rest frame
    # I do it using scalar product of A to p1
    vecTildeAsq = mprod(amp,p1)*mprod(p1,ampc)/mJpsi^2 - mprod(amp,ampc)
    # attach muon current avaraged over muons angles
    ksq = λ(mJpsi,mμ^2,mμ^2)/(4*mJpsi^2)
    return 2*vecTildeAsq*(mJpsi^2/4-ksq/3)
end

amplsp_analytic (generic function with 1 method)

In [308]:
plot(0.5:0.03:2.5, -.99:0.03:0.99, 
    (s,zs)->(sqrt(λ(s,mπ^2,mK^2)*λ(s,mJpsi^2,mB^2))/s*
amplsp_analytic(s,zs,[[0,0,1]])), t=:heatmap)

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 0.5 
 
 
 1.0 
 
 
 1.5 
 
 
 2.0 
 
 
 -1.0 
 
 
 -0.5 
 
 
 0.0 
 
 
 0.5 
 
 
 1.0 
 
 
<image width="475" height="364" xlink:href="data:;base64,
iVBORw0KGgoAAAANSUhEUgAAAdsAAAFsCAYAAACEtRP5AAAgAElEQVR4nO3dyXJjSZJmYb0DAA4e
Q0ZmbUpapPuJ+kn6xXvTXSWVEUFiurXgWo+FKKgg3PN8W3WzewECVED4u9r0v3/6P1uA//U6p7X/
+e2U1v7Hy+9p7d9//s+09o9f/iOt/fqP/5fWXv8tX7f/R3695bdjWpv+/pLWtl//ltYiIi6//D2t
Xb/9lq97+TW/5lO+Lvb5unn3U1pb1vwxzvNTWpvmfb5uWtNaTFNeG9nyl+q2XaB2TmvXa/7z59pb
Xrv8mdbiDLVT/p6ZznktImI6/ZHXjvk1p1Nem99h3TtcD9flzxvV4v0davnPaXvLXxcRERtsG+/5
a/V6zF/j23teux53ae3ynr+nLu+wjvaE2plqJ3gPR8T5VFt7Oue1M9Ro3elSW3e8LLAn1K55jfbM
O6kkSfoUNltJkprZbCVJamazlSSpmc1WkqRm6+l6xX/wds378Z9nquVptd/fD2nt9S1PwB7+6zWt
Lbs8cTrNkEjc/pnveaHEZZ7EjohY/szXzt/ydPT8+nNa257/b1q7Hr7l63bPae28y5/vWPLatuTJ
ycA08ujzHaWVITgPaeS4wmvjCj9HSCPHJa/NZ0i4w7rpnEdjpxPFZiOmE6R8j7DvsbYuaE9KDh/h
+aZU8RGS6PTUHDn9jqliqNE6ThznNVxH6V+4F1wHqeFRGvlE+0IilxLHZ0wVU3L4voljWneGdX6z
lSSpmc1WkqRmNltJkprZbCVJarYeNw5IvUPu5M9L3qv/gD8i/37KwzW/v+Vhnv0uD1csa36j00yP
MQ9QbNd8RN4CwZKIiOnt/6e1+XcIT73k4antGcZHPkGYaZ8/p9suD6ttax6C2BYKQeU/+2FACsc5
UkAKahACnK7FYNUlr8U5f51OWINA1okDebj2CKEs2pdqx/zxb0d4TuE2N7jcdsxfNxuEebYjvBYj
4gprKbDEYSYILBXDTLhnMeiEYSWoRdRDUBguooBUNehUDUFhLX8tnrf895ffbCVJamazlSSpmc1W
kqRmNltJkprZbCVJarYegw9XPkKQ9+2SJ6/+gPFav0Pq7p/HPB27h6TyshTTyJAeO0DaepRynGFk
3fwHjLp7htpTPlpyOsD4xD2kind5LTCNDI9/hs9wN6WRAaWRKXGPSWVYd4H3DSSVOcUMe55h3Wjt
CZLDJ3jeKB2MNUj4w++F7UQ1SA1TGnkwdpDWUsqXxjVeII1bThUXk8MXeL55dCL/fsM0cjFVTHtW
RyRScphHMlLiGGpwPb/ZSpLUzGYrSVIzm60kSc1stpIkNVtPo4DUJQ9QYEAK/sD8J/wB/Z/HPOiz
h/NVFzizdpooPJM/hiv8sXs/CF6s7/k4x/VPCkHlyZP5AKP+9hCgoaNnIR8VEIKaFvicRrVhQIrL
JTjKsTbmEWsXqsG5rPRWHOSjqms3OJMaw0zldRR0KgakMJBUH9dIQacrPEYMSBUDS7gOfp9ecKxi
LXT0cc37Bp3O0E8o6ESBJR67WAtBOa5RkqQvZLOVJKmZzVaSpGY2W0mSmtlsJUlqth5pJltEvMOo
uzdIlv0JacXfIXX3tOTx2D2MclxgJCOlkbdiGplGq0VE7N/zCPAF0sjLU366NqWR5z0klVdIaq/w
vMEIzKAafYRrmtaIimfO07qAwDHVNkhVxhXGHNK6UZ3GjsJ7mBK3nDiu7Vmv1dK4o30pAYyJY7gm
7knriunfCyZ8a7WIQSK3YbQiXq+45wV+91cTx45rlCTpC9lsJUlqZrOVJKmZzVaSpGbraeKA1BHG
0r3BH6b/xLNu8x7/tOTBg3XOQ0cLhKAmSLpUA1I0Wi0i4nzMg167t/xxrBCCWiAEtewgPLWD81Ub
wlPTDMkiGp35V+qfDX7+XIMSvG42CEEFrbshIIUhqHuvo6AT7Hktho5oXUQ9lIRBJ/g54uMojius
h6Bqe0bUA0u8rjoi8fNHK9LPEANSjmuUJOnr2GwlSWpms5UkqZnNVpKkZjZbSZKarafhuMY8dXqE
Q4sxjQwpuAOkfHcwkrGcRobTyjGNPEg5niCNvD/mIxl3b7U08rrmieMZEsdYo8TxAkllSiPDzzCC
R2vyQqhVA86QLKQUOyaOcR2lmAdpZLgmJoeLyVlKP9P7hvak+7zg9WqJ4uE1y6ni2p4taeTivYwO
j+e1n58cvsDrm/asjmTEdbinaWRJkr6MzVaSpGY2W0mSmtlsJUlqtp6mPKwTEXHc8uDN+zUPAdFZ
t3+c8z8iH+a8/++mPDxV/dRwhWQNhQBOg+DFEwS9ThQCg6AThaCothRDUDOFoKA2T8XwVASOayyH
p6qqASk8I5fWUQiID/qtjojEgNS9a9XQVTEE1LUvBZbq4alqsKi4J7wWb1lbDSVh7c5Bp+qefrOV
JKmZzVaSpGY2W0mSmtlsJUlqtp7jHf/BMSgglQdh3iHsQNOl9hCeWqc8eDBzfiRFQZdbJqycYLrW
/gzTpSgEtVAtDzotxRDUMtOZtRCQghDUKOREdVyLP/9isIrOs6Vl5WBVrTZcS9OlKFhVPOv5CgGZ
6r1ci/dyS0DqQpOJOoJVxevVpyvV9rxlLa6rBqRagk7V+0xLfrOVJKmbzVaSpGY2W0mSmtlsJUlq
ZrOVJKnZeo7BuEYY5/i+PeU1SOu+URoZYsUrjHKcqmlkqFXPQoyIOMI4x8M5TxXvIDm8K6aRMamM
qeJa4hjHNXalkYvodYNjF4vKaeSufRvGR2KKubhndR2lXz/W1lKn9RRz7TF2jCusJoMjRs/546SR
e2ppyfNsJUn6SjZbSZKa2WwlSWpms5Ukqdl63k74D44QoHqHUY5vVxhXCAGpFRIrC9Sm4ueGK/yx
uzoGLCLiCAGx4wLn2cKIRAxPwboVgk40kpHDU7WA1NwUkLr7WbdFo7GL1XUbzKukoBeNQSyHpzAg
RaGjYkCqeL1b9u0IOuH1Ou6TzvIePm+fHzyqPo57h6Cqe/rNVpKkZjZbSZKa2WwlSWpms5UkqZnN
VpKkZuslOI18mvLD5d8hyfx+zQ9If4OD5TGNDB8NJpzXSEm+fBUly050MxFx2PKU7wnGq+3mPMW8
u+Qp37WaRobkMKWRF0j/VpPKERETDCaESZ4RlFSmYYewZ3ECKOIRkF3jGmEdppiLIxmLe5ZTzDel
ke99P5+fVOZaWvqSNHJPijkttaSR+THk6/xmK0lSM5utJEnNbLaSJDWz2UqS1Gy9DsY1niEgRWfd
vm37tLaHUYY7Og/wTOMaQUNg4TwIEJzgcewheLSHcBEFnVYIVq0QHuIQVDEgVayN6jiuESJElJ3D
8NQDoSDTcG1HQAquVx4BWdyzer0IDgl1BKQ6HmO5BvfSF5Bq2DPfsiVYVV3nN1tJkprZbCVJamaz
lSSpmc1WkqRmNltJkpqtl0Ea+QSHx1MamUY5vm15cnaBg+Vn+GzA4xpzPLIsX3eGFHNExAFmC57m
fO0R0sE7WLdCcniFhDOlijsSx7ekkWdMHFNSOccvm9qeHW7JTGOSGZPKtT070s+UAK3uGcFJVh5J
WVx35z3zIa63jrks7lv8fVseSdny3OTrHNcoSdIXstlKktTMZitJUjObrSRJzdbrdsZ/cNnyENRp
ektrxykf1/i2rWltgbNe8TxTOCOXxzXClkv9v

In [281]:
# S-wave
function projected_dalitz(s,par)
    sqrt(λ(s,mπ^2,mK^2)*λ(s,mJpsi^2,mB^2))/s*
        quadgk(zs->amplsp_analytic(s,zs,par),
                -0.999,0.999)[1]
end

plot(hists)
norm1 = quadgk(s->projected_dalitz(s,[[1,0,0]]),((mπ+mK)^2+0.01),((mB-mJpsi)^2-0.001))[1]
plot!(s->projected_dalitz(s,[[1,0,0]])/norm1, 
    ((mπ+mK)^2+0.01):0.02:((mB-mJpsi)^2-0.001),
    lab="j=1, l=0")
# D-wave
norm2 = quadgk(s->projected_dalitz(s,[[0,0,0.01]]),((mπ+mK)^2+0.01),((mB-mJpsi)^2-0.001))[1]
plot!(s->projected_dalitz(s,[[0,0,0.01]])/norm2,
    ((mπ+mK)^2+0.01):0.02:((mB-mJpsi)^2-0.001),
    lab="j=1, l=2")

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 1 
 
 
 2 
 
 
 3 
 
 
 4 
 
 
 5 
 
 
 0.0 
 
 
 0.1 
 
 
 0.2 
 
 
 0.3 
 
 
 0.4 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<polyline clip-path="url(#clip1902)" style="stroke:#3da44d; stroke-width:1; stroke-opacity:1; fill:none" points="
 53.0531,374.985 55.2075,371.693 57.362,367.323 59.5164,362.338 61.6709,356.983 63.8253,351.411 65.9798,345.727 68.1343,340.002 70.2887,334.291 72.4432,328.631 
 74.5976,323.051 76.7521,317.569 78.9065,312.2 81.061,306.954 83.2154,301.837 85.3699,296.852 87.5243,292.003 89.6788,287.289 91.8332,282.71 93.9877,278.264 
 96.1421,273.949 98.2966,269.763 100.451,265.702 102.606,261.764 104.76,257.944 106.914,254.241 109.069,250.649 111.223,247.167 113.378,243.789 115.532,240.514 
 117.687,237.337 119.841,234.256 121.996,231.267 124.15,228.368 126.304,225.554 128.459,222.824 130.613,220.175 132.768,217.604 134.922,215.108 137.077,212.686 
 139.231,210.334 141.386,208.05 143.54,205.833 145.695,203.679 147.849,201.588 150.003,199.557 152.158,197.585 154.312,195.669 156.467,193.808 158.621,192 
 160.776,190.244 162.93,188.538 165.085,186.881 167.239,185.272 169.394,183.708 171.548,182.19 173.702,180.715 175.857,179.282 178.011,177.891 180.166,176.54 
 182.32,175.228 184.475,173.955 186.629,172.718 188.784,171.519 190.938,170.354 193.093,169.225 195.247,168.129 197.401,167.066 199.556,166.036 201.71,165.037 
 203.865,164.069 206.019,163.132 208.174,162.224 210.328,161.345 212.483,160.495 214.637,159.672 216.792,158.877 218.946,158.108 221.1,157.366 223.255,156.65 
 225.409,155.959 227.564,155.293 229.718,154.651 231.873,154.034 234.027,153.44 236.182,152.869 238.336,152.322 240.49,151.797 242.645,151.294 244.799,150.813 
 246.954,150.353 249.108,149.915 251.263,149.498 253.417,149.101 255.572,148.725 257.726,148.369 259.881,148.034 262.035,147.717 264.189,147.42 266.344,147.143 
 268.498,146.884 270.653,146.645 272.807,146.424 274.962,146.221 277.116,146.037 279.271,145.87 281.425,145.722 283.58,145.592 285.734,145.479 287.888,145.384 
 290.043,145.306 292.197,145.246 294.352,145.202 296.506,145.176 298.661,145.167 300.815,145.174 302.97,145.199 305.124,145.24 307.279,145.297 309.433,145.371 
 311.587,145.462 313.742,145.569 315.896,145.692 318.051,145.832 320.205,145.988 322.36,146.16 324.514,146.349 326.669,146.553 328.823,146.774 330.978,147.01 
 333.132,147.263 335.286,147.532 337.441,147.817 339.595,148.119 341.75,148.436 343.904,148.769 346.059,149.119 348.213,149.485 350.368,149.867 352.522,150.265 
 354.677,150.68 356.831,151.111 358.985,151.558 361.14,152.022 363.294,152.502 365.449,152.999 367.603,153.512 369.758,154.042 371.912,154.589 374.067,155.153 
 376.221,155.734 378.375,156.332 380.53,156.947 382.684,157.58 384.839,158.23 386.993,158.897 389.148,159.583 391.302,160.286 393.457,161.007 395.611,161.747 
 397.766,162.505 399.92,163.281 402.074,164.076 404.229,164.89 406.383,165.723 408.538,166.576 410.692,167.448 412.847,168.34 415.001,169.252 417.156,170.184 
 419.31,171.137 421.465,172.111 423.619,173.106 425.773,174.122 427.928,175.16 430.082,176.221 432.237,177.303 434.391,178.409 436.546,179.537 438.7,180.69 
 440.855,181.866 443.009,183.067 445.1

## Example of the interest
Consider two resonances:
 * $K^*(1410)$, $m = 1.414\,$GeV, $\Gamma = 232\,$GeV
 * $K_2^*(1430)$, $m = 1.432\,$GeV, $\Gamma = 109\,$GeV

In [294]:
mKs = 1.414; ΓKs = 0.232;
mKs2 = 1.432; ΓKs2 = 0.109;
function bwDalitz(s,msq,Γ)
    phsp = sqrt(λ(s,mπ^2,mK^2)*λ(s,mJpsi^2,mB^2))/s
    phsp*abs(bw(s,msq,Γ))^2
end
plot(s->bwDalitz(s,mKs^2,ΓKs), (mπ+mK)^2:0.01:(mB-mJpsi)^2)
plot!(s->bwDalitz(s,mKs2^2,ΓKs2), (mπ+mK)^2:0.01:(mB-mJpsi)^2)

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 1 
 
 
 2 
 
 
 3 
 
 
 4 
 
 
 5 
 
 
 0 
 
 
 100 
 
 
 200 
 
 
 300 
 
 
 400 
 
 
 500 
 
<polyline clip-path="url(#clip4502)" style="stroke:#009af9; stroke-width:1; stroke-opacity:1; fill:none" points="
 35.5169,375.813 36.6639,375.182 37.8108,374.917 38.9578,374.711 40.1047,374.536 41.2516,374.379 42.3986,374.236 43.5455,374.102 44.6925,373.976 45.8394,373.856 
 46.9864,373.741 48.1333,373.63 49.2802,373.521 50.4272,373.416 51.5741,373.313 52.7211,373.212 53.868,373.112 55.015,373.014 56.1619,372.916 57.3088,372.82 
 58.4558,372.724 59.6027,372.628 60.7497,372.533 61.8966,372.438 63.0436,372.343 64.1905,372.248 65.3374,372.153 66.4844,372.057 67.6313,371.961 68.7783,371.865 
 69.9252,371.767 71.0722,371.669 72.2191,371.571 73.366,371.471 74.513,371.371 75.6599,371.269 76.8069,371.166 77.9538,371.062 79.1008,370.957 80.2477,370.85 
 81.3947,370.742 82.5416,370.632 83.6885,370.521 84.8355,370.408 85.9824,370.293 87.1294,370.176 88.2763,370.058 89.4233,369.937 90.5702,369.814 91.7171,369.689 
 92.8641,369.562 94.011,369.432 95.158,369.299 96.3049,369.164 97.4519,369.027 98.5988,368.886 99.7457,368.742 100.893,368.596 102.04,368.446 103.187,368.293 
 104.334,368.137 105.48,367.977 106.627,367.813 107.774,367.645 108.921,367.474 110.068,367.298 111.215,367.118 112.362,366.934 113.509,366.745 114.656,366.551 
 115.803,366.353 116.95,366.149 118.097,365.94 119.244,365.725 120.391,365.505 121.538,365.279 122.685,365.046 123.832,364.807 124.978,364.562 126.125,364.309 
 127.272,364.049 128.419,363.782 129.566,363.507 130.713,363.224 131.86,362.933 133.007,362.633 134.154,362.323 135.301,362.005 136.448,361.676 137.595,361.338 
 138.742,360.989 139.889,360.629 141.036,360.257 142.183,359.873 143.33,359.477 144.477,359.069 145.623,358.646 146.77,358.21 147.917,357.76 149.064,357.294 
 150.211,356.812 151.358,356.314 152.505,355.799 153.652,355.267 154.799,354.716 155.946,354.145 157.093,353.555 158.24,352.944 159.387,352.312 160.534,351.657 
 161.681,350.979 162.828,350.277 163.975,349.55 165.122,348.797 166.268,348.017 167.415,347.21 168.562,346.373 169.709,345.507 170.856,344.61 172.003,343.682 
 173.15,342.721 174.297,341.727 175.444,340.698 176.591,339.634 177.738,338.535 178.885,337.399 180.032,336.227 181.179,335.017 182.326,333.77 183.473,332.485 
 184.62,331.163 185.766,329.803 186.913,328.408 188.06,326.978 189.207,325.514 190.354,324.018 191.501,322.492 192.648,320.939 193.795,319.363 194.942,317.767 
 196.089,316.157 197.236,314.536 198.383,312.911 199.53,311.29 200.677,309.678 201.824,308.084 202.971,306.517 204.118,304.986 205.265,303.501 206.411,302.072 
 207.558,300.711 208.705,299.427 209.852,298.232 210.999,297.136 212.146,296.15 213.293,295.284 214.44,294.546 215.587,293.945 216.734,293.486 217.881,293.176 
 219.028,293.019 220.175,293.015 221.322,293.167 222.469,293.472 223.616,293.927 224.763,294.529 225.909,295.27 227.056,296.144 228.203,297.141 229.35,298.254 
 230.497,299.47 231.644,300.78 232.791,302.172 233.938,303.636 235.085,305.159 236.232,306.733 237.379,308.346 238.526,309.988 239.673,311.65 240.82,313.324 
 241.967,315.002 243.114,316.676 244.261,318.341 245.408,319.99 246.554,321.619 247.701,323.223 248.848,324.799 249.995,326.344 251.142,327.856 252.289,329.331 
 253.436,330.77 254.583,332.169 255.73,333.53 256.877,334.851 258.024,336.132 259.171,337.374 260.318,338.576 261.465,339.738 262.612,340.862 263.759,341.948 
 264.906,342.997 266.053,344.01 267.199,344.987 268.346,345.929 269.493,346.839 270.64,347.715 271.787,348.561 272.934,349.376 274.081,350.162 275.228,350.919 
 276.375,351.649 277.522,352.353 278.669,353.032 279.816,353.686 280.963,354.317 282.11,354.925 283.257,355.511 284.404,356.077 285.551,356.623 286.697,357.149 
 287.844,357.657 288.991,358.148 290.138,358.621 291.285,359.078 292.432,359.519 293.579,359.946 294.726,360.357 295.873,360.755 297.02,361.14 2

In [300]:
plot(s->projected_dalitz(s,[[],[0,0,1]]), 
    ((mπ+mK)^2+0.01):0.02:((mB-mJpsi)^2-0.001),
    lab="j=2, l=1")

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 1 
 
 
 2 
 
 
 3 
 
 
 4 
 
 
 5 
 
 
 200000 
 
 
 400000 
 
 
 600000 
 
 
 800000 
 
<polyline clip-path="url(#clip5702)" style="stroke:#009af9; stroke-width:1; stroke-opacity:1; fill:none" points="
 55.2858,375.734 57.5011,374.676 59.7164,372.057 61.9317,367.789 64.1471,361.949 66.3624,354.688 68.5777,346.188 70.793,336.642 73.0083,326.235 75.2236,315.141 
 77.4389,303.521 79.6543,291.518 81.8696,279.259 84.0849,266.856 86.3002,254.406 88.5155,241.992 90.7308,229.686 92.9461,217.549 95.1614,205.632 97.3768,193.977 
 99.5921,182.62 101.807,171.589 104.023,160.908 106.238,150.595 108.453,140.663 110.669,131.122 112.884,121.981 115.099,113.243 117.315,104.909 119.53,96.9805 
 121.745,89.4549 123.961,82.3292 126.176,75.5989 128.391,69.2585 130.606,63.3017 132.822,57.7215 135.037,52.5103 137.252,47.6599 139.468,43.1621 141.683,39.0081 
 143.898,35.189 146.114,31.6957 148.329,28.5192 150.544,25.6501 152.76,23.0795 154.975,20.798 157.19,18.7965 159.406,17.0662 161.621,15.598 163.836,14.3831 
 166.051,13.4129 168.267,12.6789 170.482,12.1727 172.697,11.8861 174.913,11.811 177.128,11.9397 179.343,12.2643 181.559,12.7775 183.774,13.4718 185.989,14.3401 
 188.205,15.3755 190.42,16.571 192.635,17.9203 194.851,19.4166 197.066,21.0539 199.281,22.826 201.496,24.7268 203.712,26.7508 205.927,28.8923 208.142,31.1457 
 210.358,33.5058 212.573,35.9675 214.788,38.5258 217.004,41.1757 219.219,43.9127 221.434,46.732 223.65,49.6293 225.865,52.6003 228.08,55.6408 230.296,58.7467 
 232.511,61.9141 234.726,65.1391 236.941,68.4181 239.157,71.7475 241.372,75.1239 243.587,78.5437 245.803,82.0038 248.018,85.501 250.233,89.0323 252.449,92.5946 
 254.664,96.1851 256.879,99.8009 259.095,103.439 261.31,107.098 263.525,110.774 265.741,114.465 267.956,118.169 270.171,121.884 272.386,125.606 274.602,129.335 
 276.817,133.067 279.032,136.802 281.248,140.536 283.463,144.269 285.678,147.998 287.894,151.722 290.109,155.438 292.324,159.146 294.54,162.843 296.755,166.528 
 298.97,170.201 301.186,173.858 303.401,177.5 305.616,181.124 307.831,184.729 310.047,188.314 312.262,191.879 314.477,195.421 316.693,198.94 318.908,202.435 
 321.123,205.904 323.339,209.347 325.554,212.764 327.769,216.152 329.985,219.511 332.2,222.841 334.415,226.14 336.631,229.408 338.846,232.644 341.061,235.848 
 343.276,239.018 345.492,242.155 347.707,245.258 349.922,248.326 352.138,251.358 354.353,254.355 356.568,257.315 358.784,260.239 360.999,263.126 363.214,265.975 
 365.43,268.787 367.645,271.56 369.86,274.295 372.076,276.992 374.291,279.649 376.506,282.268 378.721,284.847 380.937,287.387 383.152,289.887 385.367,292.348 
 387.583,294.768 389.798,297.149 392.013,299.489 394.229,301.79 396.444,304.05 398.659,306.27 400.875,308.45 403.09,310.59 405.305,312.69 407.521,314.749 
 409.736,316.769 411.951,318.749 414.166,320.689 416.382,322.589 418.597,324.45 420.812,326.271 423.028,328.053 425.243,329.796 427.458,331.5 429.674,333.166 
 431.889,334.793 434.104,336.381 436.32,337.932 438.535,339.444 440.75,340.92 442.966,342.357 445.181,343.758 447.396,345.123 449.612,346.45 451.827,347.742 
 454.042,348.998 456.257,350.219 458.473,351.404 460.688,352.555 462.903,353.671 465.119,354.753 467.334,355.802 469.549,356.817 471.765,357.8 473.98,358.75 
 476.195,359.668 478.411,360.555 480.626,361.41 482.841,362.235 485.057,363.029 487.272,363.793 489.487,364.528 491.702,365.234 493.918,365.912 496.133,366.561 
 498.348,367.183 500.564,367.778 502.779,368.347 504.994,368.889 507.21,369.406 509.425,369.898 511.64,370.366 513.856,370.809 516.071,371.229 518.286,371.626 
 520.502,372.001 522.717,372.354 524.932,372.686 527.147,372.997 529.363,373.287 531.578,373.558 533.793,373.811 536.009,374.044 538.224,374.26 540.439,374.459 
 542.655,374.641 544.87,374.807 547.085,374.957 549.301,375.093 551.516,375.214 553.731,375.322 555.947,375.417 558.162,375.5 560.377,375.571 562.592,375.631 
 564.8

In [315]:
plot(0.5:0.03:2.5, -.99:0.03:0.99,
    (s,zs)->(sqrt(λ(s,mπ^2,mK^2)*λ(s,mJpsi^2,mB^2))/s*
amplsp_analytic(s,zs,[[],[0,0,1]])), t=:heatmap)

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 0.5 
 
 
 1.0 
 
 
 1.5 
 
 
 2.0 
 
 
 -1.0 
 
 
 -0.5 
 
 
 0.0 
 
 
 0.5 
 
 
 1.0 
 
 
<image width="475" height="364" xlink:href="data:;base64,
iVBORw0KGgoAAAANSUhEUgAAAdsAAAFsCAYAAACEtRP5AAAgAElEQVR4nO3d23Ik2ZFeYY9DHgBU
FVndHM4MpQfQ+z+NbvQGMomHJhtAZkboAryRmXztlicchRlb363X3hF5wkYa/nKfvp7/2x7g9+t/
SWv/tv3XtPbvy9d83cOc1v70sOV7PryktX99+Eda++PXv6a1n77ltW/f/5zWnv6Q1yIijj//Ja0t
Pz2ntfmnY1rbv/8+rW2/+ymt3b7+nK97ytft5+957ZSvm47f0tqyPKW1t/o5rc1z/txM05pvOk15
Dd79e+TvxX27pLVth9otfw9v26957Zq/vyMi4vL3tDRd/gbrfklL82t+zeklv94MtemZanC9X6H2
nH+e4tf8+Y6I2J+veS1/OWJ/XtLa9nLIa8957fZ8glr+3r++1GoXuM/La77urQ5rL1C75p/T12u+
7uUC6260Z157oXVb/vq+3qC25WdbXpEkSe/Cw1aSpGYetpIkNfOwlSSpmYetJEnN1usGSb6I+HXP
k4x/mfMk7+mWJ8vW54f8gnt+/l+2PKn6DAmxXy55su7P/8jTsd//mieqv/7PPI0bEfHlS566fPia
105QO3zLE5nL0/9Ia+vTf09r02Naiukxfw3jnCcn91NeiyPUImI/QBp7hZTnAmnkOX9P7dP7/745
7XmKedluaW2FWlzz1GxERNzy+kRrr3lyerq85ususOdrvifVIMQdO9zKDoHjDVKsb/vmP1O213zt
Bmlcqt3KtfxerrDuCsngKzw3lCiOiLhCyvcKKd8L/Jy+3vLP4gbnAtYC/idC0TTl/4Vhptq734kk
Sfq/eNhKktTMw1aSpGYetpIkNVtvG7czu2x5YOeX5X+nteOU/4F9ueZ/tN4jD9Bc4Q/hL7c8WPPL
Nf+j/F8hCPC/IMj17Ze8JWFExLc/573ens55KO0RameonaCV5QHWHR7y5MlyyvdcznmaZT5CO8pj
3h4wIiLWPCQ0UW2Bvov5yx+UjypnpyiTgc1RQZ65etuW2k5C7or23SCwEtCWbr9C7ZYHkna43g6B
nO2Sv8A7fPYjIrbivrTuBsGjG9wPr6vtyUEmuBcIOb3tS0EnqsHjgBaJN3i/bTucJ1RLK4w+3lTz
m60kSc08bCVJauZhK0lSMw9bSZKaedhKktRs3agPWkRcYKD1y5S3a/wrDPqe4YzfoQvcbc/3pGG/
/4B03N8u+b18O+RJ5a+D4cpfIMn89Pc8yft4gNoxTwc/HPLX8QR7Ho957QDrqrXlwG0HqT5DGnle
KKlM6/I47jTnecVphhgvtGzj7nHVfORgY9q2muTcqAaf7+o6SirDZx8T1YP6Bj83qutuxRrvmd8L
Xg+eb1oXEXGF55zTyPk1r3Q/2K6xljjeO1o5whX9ZitJUjMPW0mSmnnYSpLUzMNWkqRm606JpIig
ebevUz5fdZ7+AjUIO8AfmG+3L/m9bHmbxxf4g/0/oHUktXl8gmBVRMTTmrcle1rycNUjhIAeljx0
dYZ15yV/jc9rXjuuEKyidXC9A6yLiFggzLRCbYGgE4Wnlvn9w1PzVFtHwSqaoTnW0ZgOrgaBlZZA
FqzbIHQz2pfWttQoBFS8z2oLRApdDddWr4k1ePz0+uP7LS1hDcFb32+2kiQ187CVJKmZh60kSc08
bCVJauZhK0lSs3UPToduMFz+Cmnk1ylPpE0zJMQgdXmDSdeXDZLKlzz9+7zlqeFfIZH3y8JJzcc1
f/xUe1iolrePfIDE7RlStSdYd4SWhEdah3vSJPOIAyWZYe0KjxETzrDnAo9/phqliqvrBmlkqk8N
LSJ5T7haMeWJCec71mECupxypVoxcVxch20Oi8ngt/uBa1YTx9iSsVqjVo7VGoCi32wlSWrmYStJ
UjMPW0mSmnnYSpLUbN13DqxQgOoK4alpyoNHE53xGJ7KwyUbBE+u+2Nae71Bm8ctDyS9QEuyiIjn
W/44qEXkw5o/N2e45Hmmdflf7U8UnoLnlNZRsOpAc2AHdQpI0boVaxSegtBVsUatHOdiC8i3ei1c
dU8o673XVeFc0sGt4LzTagiKwjXV0FW5XWEtPETrIu4IZVWDVbSu+JyW5+BSm0d47f1mK0lSMw9b
SZKaedhKktTMw1aSpGYetpIkNVsDWiBGROyRpzX3PR8ufoOh8xc844vTdyEBuUGS87blj++6n9Pa
5ZonlSMiXuc8OnyBJPMrpPyeMY2crztBa0msQcL5BAnnI6RqKak8qlcTxwd4b3BSGRLH8J6qppgp
cbwMEr7V9pGYRoZMZrU9JO1ZnFXfhltLFoeSF5PKHa0jq0nd8dpiGrm6Z8tg+fdv5eg3W0mSmnnY
SpLUzMNWkqRmHraSJDVbxwMmISAFtY3CU/trWpt2mIMLfySfKF1BiQ0I8+wb1AZ94DZo9XiDpVd4
jBQguEJA6gLXu0C44ALhqVd4bk4YkOLgxQFCWdXw1AHuBwNSEAKioFN1HYWuRgEpWsttIIttHsvh
KQpk1Xx0e8iIO+brFvest4CsXW8UkKquvcH3u2qYqR6CSkvl541qfrOVJKmZh60kSc08bCVJauZh
K0lSs3UU9OE/6eehDJqTu+35jFwOVuUzcq+Rh67mKQ9dzQGBrCn/XWQeBAgosEV5jmnLHyP9boTB
i2K3G+r1xKGEfN1tEIO5wSOhYBnPu4SAFKyjzlMLXY/WQSBtgXAYBaDe6vBehRek2l2qGqzigFQt
PEX5x+FA2xYd3aWq66qf/XoHqY6uVR0za+kxci1nBylJkn4gD1tJkpp52EqS1MzDVpKkZh62kiQ1
o+jrP3HONa9QfI5SzLWE8z5R+jmv3aY8GU21K7SqjIi4wjWv8DvOFR4/JWAx5ZpWIq7UyRJidzha
GOKhE0WVY5DihlaP2LEPHsdOjwSegH2DZDito0Q5PDcbR25xZjO1evzoNDK3eUxLQT9rBk8Nwvm6
g5WZ6o4tiWO83h3tGqH20W0Xy4njD74Xv9lKktTMw1aSpGYetpIkNfOwlSSp2W8ISHWoha54x+rs
2fe/3j33w3fasGc1XIF71mqjerVFJGTHYocLcvACQkC4LndHzofnOTeotqWjoA+3a6RwHPU/hZv5
bf/gXX2mObjVma0RHDzqaNeIn+9PFIKitpJ+s5UkqZmHrSRJzTxsJUlq5mErSVIzD1tJkpr9hjRy
LT3H6cj8jJ+KtZlqMFh7gmaGuOfg95R5r93PDM8b1qhFItbSEr+C1OaR9hy8nfB+yoPHa2nV+sDy
H4FuqJp+L16NkqzUV7O6ju7mB8yOx0tWWzLSlp8sjVxN8laH2X+qgfTwGPxmK0lSMw9bSZKaedhK
ktTMw1aSpGbrqM0btknDsxpCSVMeSqLajLU867XEIa2tVNvzPQ97vi4i4gD3eoDnZoV0EdbgZazW
DvDy8p55SoDW3bN2hXulea50vY+eA1utRQxmyBYDaxgQK95LdU9SDXLdsy+u++CWjNw68f33jLgn
BFVrZfqZZuTe4Dm1XaMkST+Qh60kSc08bCVJauZhK0lSMw9bSZKareO8XrHtYjE5PE95ynfB2jGt
HeKU1/a8dtzzPY+QYn6r54//CM/Nac6f0xPESo9YS0txym8F1x3nPJJH6w6wblQ/zHlekdZR4pj2
pDRyvZZf7540Ml2Tk8q0Lr/e/MHtMf+zqLddp